src/agent_function.py

```diff
@@ -1,20 +1,51 @@
 """Module providing the AgentFunction class."""
 
-from dataclasses import dataclass
+from dataclasses import dataclass, field
+from typing import Tuple, Dict
 
 from action import Action
 from percept_sequence import PerceptSequence
+from chess_puzzle_data import obtain_latest_daily_puzzle
 
 @dataclass
 class AgentFunction:
-    """Class representing an agent function, which is an abstract mathematical description of an\
+    """Class representing an agent function, which is an abstract mathematical description of an
        agent's behavior that maps any given percept sequence to an action."""
+    partial_table: Dict[Tuple[str, ...], Action] = field(default_factory=dict)
 
-    def __init_subclass__(cls):
-        cls.partial_table = dict[PerceptSequence, Action]()
+    def __call__(self, percept_sequence: PerceptSequence) -> Action:
+        return self.partial_table.get(percept_sequence, "No action found")
 
-        # Convert puzzles to percept_sequence-action pairs and add to partial table
-        # ...
+    @classmethod
+    def processGame(cls, game):
+        partial_table: Dict[Tuple[str, ...], Action] = {}
 
-    def __call__(self, percept_sequence: PerceptSequence) -> Action:
-        return self.partial_table[percept_sequence]
+        board = game.board()
+        percept_sequence = []
+        node = game
+
+        while node.variations:
+            next_node = node.variation(0)
+            move = next_node.move
+            percept_sequence.append(board.san(move))
+            board.push(move)
+
+            if len(percept_sequence) > 1:  # Ensure we have a previous percept to map the action to
+                # The percept sequence up to the current move maps to the current move (action)
+                partial_table[tuple(percept_sequence[:-1])] = percept_sequence[-1]
+
+            node = next_node
+
+        # Am I doing somethign wrong here? I feel like this should not be needed
```

Comment:
I think the problem is that the partial_table data member exists per instance of AgentFunction, but processGame is a classmethod, so it's like a static function. I think this would be fixed by removing classmethod from processGame, or making partial_table a static data member (I'm not actually sure how to do that in Python). Of the two, the latter probably makes the most sense since AgentFunction should be callable without an instance ideally.